Add EASY/HYBRID/CONSERVATIVE policies

dongahn opened the pull request:
This PR adds various backfill queuing polices and queue-depth support for optimization:

- Add EASY/HYBRID/CONSERVATIVE policies
Introduce queue_policy_bf_base.hpp and queue_policy_bf_base_impl.hpp that implement core algorithms for a majority of backfillling strategies. Derive from this class to implement EASY/HYBRID/CONSERVATIVE policies.

- Add queue-depth support for backfill policies

This PR builds on PR #502 and PR #503 so please don't merge it before those PRs land. Resolve #476, #482, and #498.

dongahn commented:
It is likely a test case will fail on one of the travis config. So I will work on that before our review. It would be best if this PR is reviewed last anyway.

dongahn commented:
OK. This PR fixed all the issues including CI test failures and new/free mismatch problem that @grondo reported. This should be ready for your reviews -- But still preferably after PR #503 and #502.

dongahn commented:
Rebased to the master after PR #502 and forced a push.

SteVwonder commented on resource/modules/resource_match.cpp:
```diff
@@ -614,59 +629,56 @@ static inline bool is_existent_jobid (const resource_ctx_t *ctx, uint64_t jobid)
+
+    if ((rc = tr.remove (jobid)) < 0) {
         if (is_existent_jobid (ctx, jobid)) {
+           // When this condition arises, we will be less likley
```
```suggestion
           // When this condition arises, we will be less likely
```

I'm not sure I'm parsing the first sentence of this comment correctly.  Where does the probability/likelihood in jobid reuse come in?  I thought that was a deterministic process.

dongahn replied:
At the moment, I can't find the commit message. But a same jobid will have to be used as it often has to be reserved multiple times before it will finally be allocated.

SteVwonder commented on resource/modules/resource_match.cpp:
```diff
@@ -521,85 +528,93 @@ static int init_resource_graph (resource_ctx_t *ctx)
     job_lifecycle_t state = job_lifecycle_t::INIT;
 
-    if (id < 0 || at < 0) {
+    if (id < 0 || now < 0 || at < 0) {
```
might also be useful to check for and error out when `at < now`

dongahn replied:
Sure. Thanks.

SteVwonder commented on qmanager/policies/base/queue_policy_base_impl.hpp:
```diff
@@ -27,11 +27,73 @@
+}
+
+
+int queue_policy_base_t::set_params (const std::string &params)
```
Does this handle nested argument lists?  For example, `sched-policy="hybrid-backfill",policy-params="queue-depth=100,reservation-depth=10"`  IIUC, this will result in `reservation-depth` being a parameter to the top level rather than being a part of the `policy-params` value.

If so, I think we either need to refactor this function or drop the usage of `policy-params` and just make the policy parameters part of the scheduler parameters (e.g., `sched-policy=hybrid-backfill,reservation-depth=10,queue-depth=100`).

dongahn replied:
Good feedback. 

From the implement point of view, this shouldn't be a problem. Both the queue parameters (specified with `queue-params`) and policy parameters (`queue-policy`) are parsed into a common map and both the generic queuing policy code and derived, specific policy code will share this common parameter map. 

As far as the queue parameters and policy parameters won't have the identically named key, (which should always be the case), this shouldn't be a problem.

dongahn replied:
> As far as the queue parameters and policy parameters won't have the identically named key, (which should always be the case), this shouldn't be a problem.

To make this a bit more robust, I will use two distinct maps, one for `queue-param` and the other for `queue-policy`.

SteVwonder commented on qmanager/policies/queue_policy_hybrid_impl.hpp:
```diff
@@ -0,0 +1,82 @@
+             != queue_policy_base_impl_t::m_params.end ()) {
+            unsigned int depth = std::stoi (i->second);
+            if (depth < MAX_RESERVATION_DEPTH)
+                queue_policy_bf_base_t<reapi_type>::m_reservation_depth = depth;
```
If `depth > MAX_RESERVATION_DEPTH`, should `m_reservation_depth` be set to `MAX_RESERVATION_DEPTH` rather than remaining as the default?

Same for `MAX_QUEUE_DEPTH`.

SteVwonder replied:
This question/comment also applies to the `apply_params` of the other classes/policies in this PR.

dongahn replied:
If `depth` is invalid like `depth > MAX_RESERVATION_DEPTH`, the code currently uses a whole-sale approach of keeping the default value. 

https://github.com/flux-framework/flux-sched/blob/e8250ea29c666ef7f590f546ac6b603b28e01c83/qmanager/policies/queue_policy_hybrid_impl.hpp#L43

https://github.com/flux-framework/flux-sched/blob/e8250ea29c666ef7f590f546ac6b603b28e01c83/qmanager/policies/base/queue_policy_base.hpp#L137

Yes, I agree. If the value is larger than the max, we should set max. I will follow up.

SteVwonder commented on qmanager/policies/queue_policy_hybrid_impl.hpp:
```diff
@@ -0,0 +1,82 @@
+    int rc = -1;
+    try {
+        std::unordered_map<std::string, std::string>::const_iterator i;
+        if ((i = queue_policy_base_impl_t::m_params.find ("queue-depth"))
```
As we discussed at coffee time, I wonder if there is some way to refactor these policy classes to reduce the amount of duplicate code, in particular in these `apply_params` methods.  One thought that came to mind which is less invasive than my original suggestions is putting the `queue-depth` handling in the `apply_params` of a parent class, since all four scheduling policies handle that parameter application in the exact same way currently.  Then each child could call it's parent's `apply_params` followed by it's own handling of additional parameters.

dongahn replied:
This is a good suggestion.

> One thought that came to mind which is less invasive than my original suggestions is putting the queue-depth handling in the apply_params of a parent class, since all four scheduling policies handle that parameter application in the exact same way currently. 

One possibility:

I can add the `queue-depth` selection in the `apply_params` method of the parent. Then, in an derived class like `queue_policy_hybrid_impl.hpp`, I can invoke this base class's `apply_params` from which its `apply_params`. I haven't  see this pattern much so let me do some study to see if there is any side-effect, though.

SteVwonder commented on qmanager/policies/queue_policy_conservative_impl.hpp:
```diff
@@ -0,0 +1,76 @@
+        if ((i = queue_policy_base_impl_t::m_params.find ("queue-depth"))
+             != queue_policy_base_impl_t::m_params.end ()) {
+            unsigned int depth = std::stoi (i->second);
+            if (depth < MAX_QUEUE_DEPTH)
```
If `MAX_QUEUE_DEPTH` is greater than `MAX_RESERVATION_DEPTH` (which is currently the case), then this can result in the "conservative" policy producing non-conservative results.  Since this is already it's own class, it would be relatively easy to special-case the behavior of this policy to either limit the `queue-depth` to `MAX_RESERVATION_DEPTH` or allow `reservation-depth` to be set to `queue-depth` even if it exceeds `MAX_RESERVATION_DEPTH`.

This would make our "conservative" policy's behavior align with the textbook definition of conservative backfill.  For users that want a highly performant policy that is mostly conservative, they can always reach for the "hybrid" policy and manually specify a large reservation depth that is still smaller than the queue depth.

dongahn replied:
Ah. Excellent. Yes I will address this.

dongahn commented:
@SteVwonder: just a heads-up. I will get to this today unless something else pops up.

dongahn commented on qmanager/config/queue_system_defaults.hpp:
```diff
@@ -0,0 +1,44 @@
+    const unsigned int DEFAULT_QUEUE_DEPTH=8192;
+    const unsigned int MAX_RESERVATION_DEPTH = 4096;
+    const unsigned int HYBRID_RESERVATION_DEPTH = 64;
+} // namespace detail
```
Thanks. Interesting that SLURM's default reservation depth is 0. This is not even a backfill policy...

dongahn commented:
@SteVwonder: I rebased this PR to the current master and addressed all of your (excellent) review items. If CI turns green, this can use your second review. The only changes you will need to review are the ones marked with `[squash]`.

dongahn commented:
OK. Travis CI just turned green!

SteVwonder commented:
LGTM! Squash away and I'll push the button.

dongahn commented:
@SteVwonder: I squashed most of them but decide to keep two commits on their own. 

- https://github.com/flux-framework/flux-sched/pull/504/commits/30bb6920b02f7941a72f4941356eb8a7028e6f8d
- https://github.com/flux-framework/flux-sched/pull/504/commits/a93d02348cd6e6629d080e9f234b849543baf45a

I thought it would be reasonable to leave a trail about these changes. If you are okay with these change, this PR is ready to go in once Travis turns green.

dongahn commented:
OK. It is green now.

SteVwonder commented:
LGTM! Thanks @dongahn!
